modules/pubsub/src/main/scala/com/snowplowanalytics/snowplow/sources/pubsub/PubsubSource.scala

```diff
@@ -7,325 +7,254 @@
+    } yield res
+    Logger[F].debug("Pulling from subscription") *>
+      io.retryingOnTransientGrpcFailures
+        .flatTap { response =>
+          Logger[F].debug(s"Pulled ${response.getReceivedMessagesCount} messages")
+        }
+  }
 
-  private def consumeFromQueue[F[_]: Sync](
+  /**
+   * Modify ack deadlines if we need more time to process the messages
+   *
+   * @param config
+   *   The Source configuration
+   * @param stub
+   *   The GRPC stub on which we can issue modack requests
+   * @param refStates
+   *   A map from tokens to the data held about a batch of messages received from pubsub. This
+   *   function must update the state if it extends a deadline.
+   */
+  private def extendDeadlines[F[_]: Async](
     config: PubsubSourceConfig,
-    control: Control
-  ): Stream[F, LowLevelEvents[Chunk[AckReplyConsumer]]] =
+    stub: SubscriberStub,
+    refStates: Ref[F, Map[Unique.Token, PubsubBatchState]]
+  ): Stream[F, Nothing] =
     Stream
-      .repeatEval {
-        Sync[F].delay(control.errorRef.get).flatMap {
-          case None            => Sync[F].unit
-          case Some(throwable) => Sync[F].raiseError[Unit](throwable)
+      .eval(Sync[F].realTimeInstant)
+      .evalMap { now =>
+        val minAllowedDeadline = now.plusMillis((config.minRemainingAckDeadline.toDouble * config.durationPerAckExtension.toMillis).toLong)
+        val newDeadline        = now.plusMillis(config.durationPerAckExtension.toMillis)
+        refStates.modify { m =>
+          val toExtend = m.filter { case (_, batchState) =>
+            batchState.currentDeadline.isBefore(minAllowedDeadline)
+          }
+          val fixed = toExtend.view.map { case (k, v) =>
+            k -> v.copy(currentDeadline = newDeadline)
+          }.toMap
+          (m ++ fixed, toExtend.values.toVector)
         }
       }
-      .evalMap { _ =>
-        // Semantically block until message receive has written at least one message
-        val waitForData = Sync[F].interruptible {
-          control.phaser.awaitAdvanceInterruptibly(control.phaser.arrive())
-        }
-        Sync[F].uncancelable { poll =>
-          poll(waitForData) *> Sync[F].delay(control.queue.getAndSet(Nil))
-        }
-      }
-      .filter(_.nonEmpty) // Would happen if phaser was terminated
-      .map { list =>
-        val events         = Chunk.iterator(list.iterator.map(_.message))
-        val acks           = Chunk.iterator(list.iterator.map(_.ackReply))
-        val earliestTstamp = list.iterator.map(_.tstamp).min
-        LowLevelEvents(events, acks, Some(earliestTstamp))
-      }
-      .evalTap { case LowLevelEvents(events, _, _) =>
-        val numPermits = events.foldLeft(0) { case (numPermits, e) =>
-          numPermits + permitsFor(config, e.remaining())
-        }
-        Sync[F].delay {
-          control.semaphore.release(numPermits)
+      .evalMap { toExtend =>
+        if (toExtend.isEmpty)
```

Comment:
We have some debug/trace logging in other places - would that be also useful here for extending deadlines?

Reply:
I added a log line to `Utils` to help with debugging ack extensions.